### DISCUSSION TITLE
Resuming Iterators After Application Crashes: Retrieving 'nextLink' for Continuity

### DISCUSSION THREAD

**Discussion by luigi-fast** [microsoftgraph/msgraph-sdk-go#547] [Q&A] [answered]
Hello,

I am currently working with iterators and have noticed that each page obtained by the iterator contains a "nextLink" internally. While going through your documentation, I came across the ability to pause the iterator. However, I have a concern regarding the possibility of my application crashing. In such a scenario, I would like to know if there is a way to resume the iterator from the exact point where the crash occurred. Is it possible to obtain each "nextLink" received by the iterator during iteration for this purpose?

Thanks

**Comment by rkodev** [comment]
Hi @luigi-fast,

To answer your question, that would be yes, for example you can access the value of the next link property for example on the model `models.GroupCollectionResponse` you can access `collectionResponse.GetOdataNextLink()`. You can then invoke the url using the request adapter and construct a new page iterator for you to resume iteration.

However, its important to note, you might want to manage the `nextlink` value in a short-term cache and test its validity before attempting to resume, paged links are managed by the server and may or may not work with the next iteration request attempt.

**Reply by luigi-fast** [reply]
Can you tell me how I can achieve what you explained with this code:

```go
req := ...Drives().ByDriveId({driveId}).Items().ByDriveItemId("root").Delta()

resp, err := req.Get(ctx, config)
if err != nil {
	return err
}

iterator, err := func(resp drives.ItemItemsItemDeltaResponseable) (*msgraphcore.PageIterator[models.DriveItemable], error) {
	iter, err := msgraphcore.NewPageIterator[models.DriveItemable](
		resp,
		me.clients.MicrosoftGraphAPI.Client().GetAdapter(),
		drives.CreateItemItemsItemDeltaResponseFromDiscriminatorValue,
	)

	return iter, err
}

if err := iterator.Iterate(ctx, func(pageItem models.DriveItemable) bool {
	// process pageItem here
	return true
}); err != nil {
	return err
}

//I can get the nextLink here but it changes asynchronously for every page fetched.
//it would need a callback or something to notify for a new nextLink, so I can persist that link in case of a crash
iterator.GetOdataNextLink()
```

**Reply by rkodev** [reply, marked as answer]
you do not need a callback to track the nextlink, you can always get the value within the iteration function and store it somewhere as it will change after each successful iteration.

As for the ability to resume, here is a function you can use
```go
func tryResumeIteration(client *msgraphsdk.GraphServiceClient, nextLink string) error {

	// construct request info
	requestInfo := abstractions.NewRequestInformation()
	requestInfo.Method = abstractions.GET
	nextLinkUrl, err := url.Parse(nextLink)
	if err != nil {
		return err
	}
	requestInfo.SetUri(*nextLinkUrl)
	
	// construct error mapping
	errorMapping := abstractions.ErrorMappings{
		"4XX": odataerrors.CreateODataErrorFromDiscriminatorValue,
		"5XX": odataerrors.CreateODataErrorFromDiscriminatorValue,
	}
	
	// send request
	result, err := client.RequestAdapter.Send(context.Background(), requestInfo, models.CreateGroupCollectionResponseFromDiscriminatorValue, errorMapping)
	if err != nil {
		return err
	}

	// reconstruct error
	pageIterator, pageIteratorErr := msgraphcore.NewPageIterator[models.Groupable](result, client.GetAdapter(), models.CreateGroupCollectionResponseFromDiscriminatorValue)
	if pageIteratorErr != nil {
		return pageIteratorErr
	}
	
	// continue iteration
}

```

**Reply by luigi-fast** [reply]
thank you it solve my problem
